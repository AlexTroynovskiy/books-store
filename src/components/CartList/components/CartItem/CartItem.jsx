import React, { useCallback } from 'react';
import PropTypes from 'prop-types';

import { Container, Typography, Avatar, Box } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

import { ButtonComponent } from '../../../_shared/Button';

const useStyles = makeStyles({
  container: {
    display: 'flex',
    flexDirection: 'column',
    margin: '40px 30px',
    borderRadius: '10px',
    boxShadow: '-11px 11px 36px -2px rgba(72,72,72,0.4)',
    padding: '40px',
  },
  buttons: {
    display: 'flex',
    alignItems: 'center',
  },
  info: {
    margin: '0 15px',
  },
  infoContainer: {
    display: 'flex',
    alignItems: 'center',
    marginTop: '10px',
  },
  buttonContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    width: '550px',
  },
  img: {
    marginRight: '20px',
    width: '70px',
    height: '70px',
  },
  price: {
    color: '#484848',
    fontWeight: 'bold',
  },
  accent: {
    color: '#484848',
  },
});

export const CartItem = ({ book, removeFromCart, count, incrementCount, decrementCount }) => {
  const classes = useStyles();

  const handleIncrement = useCallback(() => {
    incrementCount(book.id);
  }, []);

  const handleDecrement = useCallback(() => {
    decrementCount(book.id);
  }, []);

  const handleRemove = useCallback(() => {
    removeFromCart(book.id);
  });

  return (
    <Container component="div" className={classes.container}>
      <Typography variant="h4" color="primary">
        {book.title}
      </Typography>
      <Typography>{book.description}</Typography>
      <Container component="div" className={classes.buttons}>
        <Container component="div" className={classes.infoContainer}>
          <Avatar src={book.img} className={classes.img}></Avatar>
          <Typography variant="h6" color="primary" className={classes.info}>
            Price for one item:
            <Box component="span" className={classes.accent}>
              {` ${book.price}`} $
            </Box>
          </Typography>
          <Typography variant="h6" color="primary" className={classes.info}>
            Total:
            <Box component="span" className={classes.price}>
              {` ${count * book.price}`} $
            </Box>
          </Typography>
        </Container>
        <Container component="div" className={classes.buttonContainer}>
          <ButtonComponent onClick={handleDecrement} color="primary" disabled={count <= 1}>
            -
          </ButtonComponent>
          <Typography variant="h6" color="primary" className={classes.info}>
            {count}
          </Typography>
          <ButtonComponent onClick={handleIncrement} color="primary">
            +
          </ButtonComponent>
          <ButtonComponent onClick={handleRemove} className={classes.remove}>
            Remove
          </ButtonComponent>
        </Container>
      </Container>
    </Container>
  );
};

CartItem.propTypes = {
  book: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    price: PropTypes.number,
    description: PropTypes.string,
    img: PropTypes.string,
    category: PropTypes.string,
  }).isRequired,
  count: PropTypes.number,
  removeFromCart: PropTypes.func,
  incrementCount: PropTypes.func,
  decrementCount: PropTypes.func,
};

CartItem.defaultProps = {
  removeFromCart: () => {},
  count: 1,
  incrementCount: () => {},
  decrementCount: () => {},
};
