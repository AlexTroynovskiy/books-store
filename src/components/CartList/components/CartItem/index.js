import { connect } from 'react-redux';

import { CartItem } from './CartItem';
import { incrementCount, decrementCount } from '../../../../redux/Cart/actions';
import { bookCount } from '../../../../components/_shared/BookCardControlers/BookCardControlersSelector';

const mapStateToProps = (state, props) => ({
  count: bookCount(state, props),
});

const mapDispatchToProps = {
  incrementCount,
  decrementCount,
};

export default connect(mapStateToProps, mapDispatchToProps)(CartItem);
