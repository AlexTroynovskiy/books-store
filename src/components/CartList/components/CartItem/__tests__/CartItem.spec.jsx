import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import CartItemConnect from '../';
import { CartItem } from '../CartItem';

describe('Component: CartItem', () => {
  const initialState = {
    book: {},
    removeFromCart: jest.fn(),
    count: 1,
    incrementCount: jest.fn(),
    decrementCount: jest.fn(),
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('Should be properly connected to the store', () => {
    const mockStore = configureStore();
    let store, wrapper;
    store = mockStore(initialState);

    beforeEach(() => {
      wrapper = shallow(
        <Provider store={store}>
          <CartItemConnect {...initialState} />
        </Provider>
      );
    });

    test('To be in DOM', () => {
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('To have nested elements', () => {
    let wrapper;

    test('To exist', () => {
      wrapper = shallow(<CartItem {...initialState} />);
      expect(wrapper.exists()).toBe(true);
    });

    test('To match snapshot', () => {
      wrapper = renderer.create(<CartItem {...initialState} />).toJSON();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
