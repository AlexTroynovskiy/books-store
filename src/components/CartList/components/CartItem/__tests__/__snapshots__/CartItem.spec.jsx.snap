// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CartItem To have nested elements To match snapshot 1`] = `
<div
  className="MuiContainer-root makeStyles-container-1 MuiContainer-maxWidthLg"
>
  <h4
    className="MuiTypography-root MuiTypography-h4 MuiTypography-colorPrimary"
  />
  <p
    className="MuiTypography-root MuiTypography-body1"
  />
  <div
    className="MuiContainer-root makeStyles-buttons-2 MuiContainer-maxWidthLg"
  >
    <div
      className="MuiContainer-root makeStyles-infoContainer-4 MuiContainer-maxWidthLg"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circle makeStyles-img-6 MuiAvatar-colorDefault"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiAvatar-fallback"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
      <h6
        className="MuiTypography-root makeStyles-info-3 MuiTypography-h6 MuiTypography-colorPrimary"
      >
        Price for one item: 
      </h6>
      <h6
        className="MuiTypography-root makeStyles-info-3 MuiTypography-h6 MuiTypography-colorPrimary"
      >
        Total: 
        1
      </h6>
    </div>
    <div
      className="MuiContainer-root makeStyles-buttonContainer-5 MuiContainer-maxWidthLg"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          -
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          +
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Remove
        </span>
      </button>
    </div>
  </div>
</div>
`;
